### DISCUSSION TITLE
GPU use on google collab

### DISCUSSION THREAD

**Discussion by Requiem8** [IBM/aihwkit#629] [Installation Q&A] [answered]
Couldn't get the packadge to work on windows, but works fine on google colab.  google colab as I understand is Linux-based maybe why. 

Google colab, however, saying that not toolkit is not using GPUs available to it. I guess it's using pytorch and CUDA which should be GPU-hardware accelerated. Is there a setting on google collab that I'm missing? 

For context I'm a 3rd year Electronic Engineering university student (so my engineering skills are better than my coding) and I'm using this as part of my final year project on memristor crossbar arrays for inference of ANNs.

**Comment by maljoras** [comment]
Note that on conda `aihwkit-gpu` is available and need to be used if you want to enable GPU computing. Regarding running on windows, it should be in principle possible, however, you need to compile it yourself as we do only provide pre-compiled packages for linux.

**Comment by kkvtran** [comment, marked as answer]
If you want to use the GPU-enabled aihwkit package in google Colab then you need to connect to the instance with GPU.  By default, it would give you the non-GPU one.

**Comment by kkvtran** [comment]
![image](https://github.com/IBM/aihwkit/assets/22841084/a6a09a27-76d7-44fd-9a56-a63492e12cce)

**Reply by Requiem8** [reply]
Hey, thanks for your help.

on the hw_aware_training tutorial it says to:
"Upload the utils directory in your runtime environment if you are using Google Colab."

It appears that this error is caused by not doing so:
ModuleNotFoundError: No module named 'aihwkit.simulator.parameters.utils'

My questions is how do I upload the utils directory?

I tried this:`
!git init `<br>`
!git remote add -f origin https://github.com/IBM/aihwkit.git`<br>`
!git config core.sparseCheckout true`<br>`
!echo "src/aihwkit/utils" >> .git/info/sparse-checkout`<br>`
!git pull origin master`<br>`
`

But I think this is a different utils file altogether anyway.

**Comment by Zhaoxian-Wu** [comment]
Hi @kkvtran, thanks for your solution before. However, when I tried the ```pip install aihwkit``` on an Colab GPU instance, I found even though the Pytorch has available CUDA, AIHWKIT doesn't. Did I miss something?
 
![image](https://github.com/user-attachments/assets/bcf758d6-784d-48e8-9eb7-28292c063564)

**Reply by kkvtran** [reply]
the cuda-enable aihwkit is a bit big for pypi.org.  We included the following instructions in the notebook for cuda-enable kit.  Hope this works for you.

# To install the gpu enabled wheel, use the commands below
#!wget https://aihwkit-gpu-demo.s3.us-east.cloud-object-storage.appdomain.cloud/aihwkit-0.9.1+cuda117-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl 
#!pip install aihwkit-0.9.1+cuda117-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl

**Reply by Zhaoxian-Wu** [reply]
Success! I really appreciate your help
